File: HOMEFEED_BLOG_OS_R2_CODEX_PACKAGE/HOMEFEED_BLOG_OS_R2/reference_core/canonical.py
"""HFBO-NFC-INT-1: restricted, cross-language canonical bytes for immutable IR.
Not RFC8785/JCS. Integer-valued JSON numbers within the JS safe range are supported;
fractions must be modeled as decimal strings. Source raw bytes retain their own hash.
"""
from __future__ import annotations
import hashlib,json,math,unicodedata
from typing import Any
MAX_SAFE_INTEGER=9007199254740991

def _text(value:str)->str:
    if any(0xD800<=ord(c)<=0xDFFF for c in value):
        raise ValueError('CANON_INVALID_UNICODE')
    return unicodedata.normalize('NFC',value)

def normalized(value:Any)->Any:
    if value is None or isinstance(value,bool):return value
    if isinstance(value,str):return _text(value)
    if isinstance(value,(int,float)):
        if not math.isfinite(value) or int(value)!=value or abs(value)>MAX_SAFE_INTEGER:
            raise ValueError('CANON_NOT_SAFE_INTEGER')
        return int(value)
    if isinstance(value,list):return [normalized(v) for v in value]
    if isinstance(value,dict):
        result={}
        for key,v in value.items():
            if not isinstance(key,str):raise ValueError('CANON_NON_STRING_KEY')
            key=_text(key)
            if key in result:raise ValueError('CANON_NORMALIZED_KEY_COLLISION')
            result[key]=normalized(v)
        return result
    raise TypeError('CANON_UNSUPPORTED_TYPE')

def canonical_bytes(value:Any)->bytes:
    return json.dumps(normalized(value),ensure_ascii=False,sort_keys=True,
                      separators=(',',':'),allow_nan=False).encode('utf-8')

def content_hash(value:Any)->str:
    return hashlib.sha256(canonical_bytes(value)).hexdigest()

def strict_json_loads(text:str)->Any:
    def pairs(items):
        result={}
        for k,v in items:
            if k in result:raise ValueError('JSON_DUPLICATE_KEY')
            result[k]=v
        return result
    return json.loads(text,object_pairs_hook=pairs,
        parse_constant=lambda x: (_ for _ in ()).throw(ValueError('JSON_NONFINITE')))

File: HOMEFEED_BLOG_OS_R2_CODEX_PACKAGE/HOMEFEED_BLOG_OS_R2/reference_core/canonical.mjs
// HFBO-NFC-INT-1; code-point key order, NOT JS's default UTF-16 key order.
import {createHash} from 'node:crypto';
import {readFileSync} from 'node:fs';
function text(s) {
  for (const c of s) { const n=c.codePointAt(0); if(n>=0xD800&&n<=0xDFFF) throw Error('CANON_INVALID_UNICODE'); }
  return s.normalize('NFC');
}
function cmp(a,b) {
  const x=Array.from(a,c=>c.codePointAt(0)),y=Array.from(b,c=>c.codePointAt(0));
  for(let i=0;i<Math.min(x.length,y.length);i++)if(x[i]!==y[i])return x[i]-y[i];
  return x.length-y.length;
}
export function canonicalString(v) {
  if(v===null)return 'null';
  if(typeof v==='boolean')return v?'true':'false';
  if(typeof v==='string')return JSON.stringify(text(v));
  if(typeof v==='number') {if(!Number.isSafeInteger(v))throw Error('CANON_NOT_SAFE_INTEGER');return String(v);}
  if(Array.isArray(v))return '['+v.map(canonicalString).join(',')+']';
  if(typeof v==='object') {
    const normalized=new Map();
    for(const [k,x] of Object.entries(v)){const nk=text(k);if(normalized.has(nk))throw Error('CANON_NORMALIZED_KEY_COLLISION');normalized.set(nk,x);}
    return '{'+[...normalized.keys()].sort(cmp).map(k=>JSON.stringify(k)+':'+canonicalString(normalized.get(k))).join(',')+'}';
  }
  throw Error('CANON_UNSUPPORTED_TYPE');
}
export function canonicalHash(v){return createHash('sha256').update(canonicalString(v),'utf8').digest('hex');}
if(process.argv[2]==='--stdin'){
 try{const value=JSON.parse(readFileSync(0,'utf8'));process.stdout.write(JSON.stringify({canonical:canonicalString(value),hash:canonicalHash(value)}));}
 catch(e){process.stdout.write(JSON.stringify({error:e.message}));process.exitCode=2;}
}
